### DISCUSSION TITLE
A way to pass an already generated GraphQLSchema to kitql and ignore modules?

### DISCUSSION THREAD

**Discussion by frankmayer** [jycouet/kitql#260] [Ideas]
Hi @jycouet 

I would like to propose some way to pass an already generated GraphQLSchema to kitql and ignore the modules with their typedefs and resolvers.
I am currently researching 
- [Cruddl](https://github.com/AEB-labs/cruddl)  => "cruddl - create a cuddly GraphQL API for your database, using the GraphQL SDL to model your schema."
- for use with the Graph Database [ArangoDB](https://github.com/arangodb/arangodb)

Cuddle would automatically create (and also migrate on changes) the needed GraphDB Collections (nodes and edges) and all needed types and resolvers automatically based on the GraphQL SDL.
The only thing that is needed is to inject that cruddl-produced schema into graphql-yoga. I have tested this on a standalone yoga server and it works great, but it would be great if it would optionally be possible with kitql.
This would mean that we need a way to inject that schema into KitQL, so it can then inject that instead of the modules and also not bother with any schema merging of the modules. 

Cruddle would be a perfect fit for people wanting to use ArangoDB and have a relatively easy (as also with KitQL) point of entry.

I might grow out of KitQL for this specific project and run a standalone Yoga server in that configuration, but I think it would still be an interesting option for people (and me :smile:) to have.

What do you think?

**Comment by jycouet** [comment]
Thanks for reaching out 👍.

It seems like a good use case. I would like to support a variety of ways of working.

I have a few questions to understand better: 
- Once you put the schema, how fields will get resolved?
- Can't this be only a single module?

**Reply by frankmayer** [reply]
I thought it would be a great fit for kitql. At the moment it only supports ArangoDB, which is a really good GraphDB, but in theory adapters could also be written for other DBs.

To your questions: 

> * Once you put the schema, how fields will get resolved?

Cuddle would automatically create (and also migrate on changes) the needed GraphDB Collections (nodes and edges) and **all needed types and resolvers automatically based on the provided (to cruddl) GraphQL SDL.**

> * Can't this be only a single module?

I am afraid not, as it doesn't produce graphql files. It creates a ready-to-consume schema with everything in it + the database collections necessary. If you look at the [cruddl](https://github.com/AEB-labs/cruddl) project in the readme (main page) you will understand how it works.
So the schema just needs to be injected and you get a lot of resolvers for everything instantly. No need to write your own resolvers and maintain on changes. You just define the SDL (with syntactic sugar for cruddl to understand relations etc...) and it just makes it happen.

**Reply by frankmayer** [reply]
I was just thinking about...
if it would be possible to somehow combine the already produced schema from cruddl with additional modules? 

That would of course require that we would somehow be able to pass that already produced schema to a module instead of creating a module by defining typedefs and resolvers. 
In a quick look, I didn't find a way to do this with [GraphQL Modules](https://the-guild.dev/graphql/modules/docs), except if it is possible to pass the main schema and somehow add additional modules after that.

**Reply by jycouet** [reply]
👋 there, 

I think that we should connect on discord to speak about it 🥳
You can DM me with `jycouet#9418`

**Reply by frankmayer** [reply]
Nice. OK, contacted you via discord. We'll talk over there

**Reply by frankmayer** [reply]
OK, I just learned from The-Guild support about their [stitching tool](https://www.the-guild.dev/graphql/tools/docs/schema-stitching/stitch-combining-schemas), which I think is the way to go, also for KitQL  :partying_face: .  I'll do some research on that and let you know. I'll do a PR on KitQL, if this works.

**Reply by jycouet** [reply]
With: 
```
const schema = project.createSchema(db);
```

You can do: 
```
export const cruddleModule = createModule({
  id: 'cruddle-module',
  typeDefs: schema.typeDefs,
  resolvers: schema.resolvers,
})
```

And have it as a module. 🤔

**Reply by frankmayer** [reply]
yep, I tried that, but it won't work. `schema` is of type GraphQLSchema... throws:
```
src/index.ts:74:22 - error TS2339: Property 'typeDefs' does not exist on type 'GraphQLSchema'.

74     typeDefs: schema.typeDefs,
                        ~~~~~~~~

src/index.ts:75:23 - error TS2339: Property 'resolvers' does not exist on type 'GraphQLSchema'.

75     resolvers: schema.resolvers
                         ~~~~~~~~~

src/index.ts:96:8 - error TS2339: Property 'listen' does not exist on type 'YogaNodeServerInstance<{ req: IncomingMessage; res: ServerResponse<IncomingMessage>; }, {}, {}>'.

96 server.listen(4000, () => {
          ~~~~~~
```
Do you know of any other way to create a module from a schema?

**Reply by jycouet** [reply]
Would you create a repo with this so that I can have a look? 
🙏 a lot

**Reply by frankmayer** [reply]
Yes I can do that. Will cook something up, tomorrow. It's a bit late now :) . I think you can also get away without installing ArangoDB. Cruddl also has an "InMemoryDB" we can use for tests. A DB is needed for the schema to be created, because of the automatic resolvers that have to be done especially for that DB.
:pray: Thank you very much

**Reply by frankmayer** [reply]
@jycouet got a repo with a scenario here. https://github.com/frankmayer/cruddl-graphql-modules-tests

**Reply by jycouet** [reply]
Thx a lot. 
I'll have a look
